sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -163,6 +263,79 @@ checkModule module_ = do
             let (presentOrigin, context) = _present origin
             withContextF present context $ checkDefDataType presentOrigin dt
 
+-- It is always invalid to keep an interface in an upgrade
+checkContinuedIfaces
+    :: Upgrading Module
+    -> HMS.HashMap LF.TypeConName (Upgrading (DefDataType, DefInterface))
+    -> TcUpgradeM ()
+checkContinuedIfaces module_ ifaces =
+    forM_ ifaces $ \upgradedDtIface ->
+        let (_dt, iface) = _present upgradedDtIface
+        in
+        withContextF present (ContextDefInterface (_present module_) iface IPWhole) $
+            throwWithContextF present $ EUpgradeError $ TriedToUpgradeIface (NM.name iface)
+
+-- This warning should run even when no upgrade target is set
+checkNewInterfacesHaveNoTemplates :: LF.Package -> TcM ()
+checkNewInterfacesHaveNoTemplates presentPkg =
+    let modules = NM.toHashMap $ packageModules presentPkg
+        templateDefined = HMS.filter (not . NM.null . moduleTemplates) modules
+        interfaceDefined = HMS.filter (not . NM.null . moduleInterfaces) modules
+        templateAndInterfaceDefined =
+            HMS.intersectionWith (,) templateDefined interfaceDefined
+    in
+    forM_ (HMS.toList templateAndInterfaceDefined) $ \(_, (module_, _)) ->
+        withContext (ContextDefModule module_) $
+            warnWithContext WShouldDefineIfacesAndTemplatesSeparately
+
+-- This warning should run even when no upgrade target is set
+checkNewInterfacesAreUnused :: LF.Package -> TcM ()
+checkNewInterfacesAreUnused presentPkg =
```

Comment:
Improve names and comments of these two functions `checkNewInterfacesHaveNoTemplates` and  `checkNewInterfacesAreUnused`

Reply:
233ce273cd